src/crawl/DeepCrawler/DeepCrawler.js

```diff
@@ -0,0 +1,161 @@
+import * as cheerio from 'cheerio';
+
+import { BaseCrawler } from '../BaseCrawler.js';
+import { SCRAPE_PAGE_PROMPT, SCRAPE_TYPE_PROMPT, ANALYZE_PAGE_PROMPT, GENERATE_SCHEMA_PROMPT } from './prompts.js';
+
+export class DeepCrawler extends BaseCrawler {
+  // TODO(pilvcc): what is `query` and what do we do with it?
+  async *run(url, query, options) {
+    const fetchOptions = options?.fetchOptions || {};
+
+    const userPrompt = options.originalPrompt;
+
+    const scrapeType = await getPromptScrapeType(userPrompt);
+    console.log(`prompt has scrape type: ${scrapeType}`);
+
+    const schema = await generateSchemaFromPrompt(userPrompt);
+    console.log('schema generated', JSON.stringify(schema, null, 2));
+
+    const seenUrls = new Set();
+    const urlStack = [url];
+
+    while (true) {
+      const latestUrl = urlStack[urlStack.length - 1];
+      console.log('processing', latestUrl);
+
+      // we've seen this url, we're just going in circles now
+      if (seenUrls.has(latestUrl)) return null;
+
+      seenUrls.add(latestUrl);
+
+      const { html } = await this.fetcher.first(latestUrl, fetchOptions);
```

Comment:
I think better to keep the doc abstraction, so do `const doc = await ...` and then reference `doc.html`. since it's a class not a dict